src/librustc_typeck/check/mod.rs

```diff
@@ -2236,7 +2236,9 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             hir_id, def_id, substs, user_self_ty, self.tag(),
         );
 
-        if !substs.is_noop() {
+        if !substs.is_noop() || user_self_ty.map_or(false, |ty| {
+            ty.has_free_regions() || ty.has_projections() || ty.has_infer_types()
```

Comment:
Is there some justification for why these checks are enough?

Reply:
Even if these checks are enough, shouldn't they be run on the pair `(substs, ty)`? i.e.

```rust
let types = (substs, ty); // yes there are TypeFoldable impls for pairs and options
if types.has_free_regions() || types.has_projections() || types.has_infer_types() {
    // ...
}
```

Reply:
Nah I've seen a comment at https://github.com/rust-lang/rust/blob/a7ae7a21624952f5bfc6aaf88e53351385b89dd4/src/librustc_typeck/check/mod.rs#L2436-L2444

Maybe move the check there it to a separate function (`needs_user_type_annotation`), while still doing the `types` trick?